### ISSUE TITLE
actionmenu: title attribute ignored on NVDA screen reader

### ISSUE THREAD

**Issue by andylincoln** [pluralsight/design-system#466]
![actionmenutestnvda_windows_firefox](https://user-images.githubusercontent.com/2366108/52510071-9aad1a00-2bc8-11e9-8b2f-5623f873bcc1.jpg)

## Bug
I used my Windows desktop with the NVDA screen reader to get an idea of how non-Mac screen readers handle our use of the `title` attribute as a non-visual label on buttons. I found that the `title` attribute's value is getting ignored, and the labels inside in this case are being read twice. This will cause confusion at worst and annoyance at best for users of NVDA. 

### Expected behavior
NVDA reads action menu with a non-visual label of "more options" to listener.

### Actual behavior
The `title` attribute is not being read aloud, so the contents of the button are being read instead. The SVG icon's `aria-label` as well as its parent wrapping `aria-label` of "More icon" are being read aloud twice as they are the contents of the `<button>` and no other way of setting the accessible name is provided.

### Steps to repro

1. Install NVDA onto a Windows computer https://www.nvaccess.org/download/
2. Go to https://app.pluralsight.com/library/ in a browser
3. Find a course from the "Continue Learning" section by tabbing to it or hovering over it with your mouse (by default NVDA will read aloud what your mouse is hovering over).
4. Focus on or hover over the action menu for this course. NVDA will read aloud the element's accessible name

### Related packages

- actionmenu v3.2.1
- Any other packages that use `title` on as an HTML attribute

### Code where this is used
[Link to code from Prime Directive using this](https://github.com/ps-dev/narwhal/blob/cb31087d1c1812385ed37c81708068cd2cfe3661/client/javascripts/routes/home/content-tile.js#L233)

### Environment

- *OS*:  Windows 10
- *Browser version*: Firefox 65.0
- *Screen Reader*: NVDA 2018.4.1

## Enhancement Request
Use `aria-label` instead of `title` as the attribute for labeling buttons that have no text. `title`s are not an acceptable replacement to visual text for setting a `<button>`'s accessible name.

### Today's behavior
`title` as a non-visual label works in ChromeVOX and VoiceOver. This is likely because `title` as label is a common enough pattern that the Blink and WebKit teams have made it an acceptable way to set an accessible name for a certain elements. Worth noting - Google and Apple both make these browsers and the screen readers that work with them. This is not standard with the W3C's suggestions with ARIA and likely will not be made universal. `aria-label` has the widest support for making non-visual labels.

### The problem
Users are getting duplicate info from icons and the button's "label" is not actually being read aloud. A screen-reader user will likely not be able to understand this menu.

### The Value Add

[WCAG AA compliance for non-visual labels](https://www.w3.org/TR/WCAG20-TECHS/ARIA14.html)

[NVDA and JAWS make up the bulk of screen reader users](https://webaim.org/projects/screenreadersurvey7/#primary). NVDA not supporting this method of labeling is a huge problem for any users of our applications that need assistive tech.

JAWS, if I recall correctly also does not support `title` as label, at least by default. We have yet to get our JAWS licenses (though those are on the way thanks to Maureen Botoman), so that will need to be tested as well to confirm.

### The Tradeoffs

`title` as label seems to be used a lot across the design system. If we are using `title` as an invisible label anywhere else and not just on actionmenu, they will also need to be updated. 

`title` is also a way to get a "tooltip", so an alternative way of providing tooltips will be needed.

_Note on Tooltips:_ I suggest checking out [Tooltips & Toggletips by Heydon Pickering](https://inclusive-components.design/tooltips-toggletips/) on his [Inclusive Components blog](https://inclusive-components.design/) for ways to make tooltips accessible.

This could potentially be a breaking change if I understand correctly, everyone using this package will need to migrate to a version that doesn't use `title`, requiring the prop to change name.

### Related literature on why `title` does not provide accessible labeling
- [HTML5 Accessibility Chops: title attribute use and abuse](https://developer.paciellogroup.com/blog/2012/01/html5-accessibility-chops-title-attribute-use-and-abuse/) by The Paciello Group
- [Using the HTML title attribute](https://developer.paciellogroup.com/blog/2013/01/using-the-html-title-attribute-updated/) by The Paciello Group
- [The Trials and tribulations of the title attribute](https://www.24a11y.com/2017/the-trials-and-tribulations-of-the-title-attribute/) by Scott O'Hara
- [HTML title attribute docs from MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/title)

**Comment by andylincoln** [comment]
I'm available for any assistance in providing a fix for this, I have a few years experience doing accessibility work. Feel free to reach out to me on Slack if needed.

**Comment by jaketrent** [comment]
Why not leave `title` on the <button> (for the visual tooltip feature) and add `aria-label` (for the screenreader support)?

**Comment by andylincoln** [comment]
@jaketrent If we can test and confirm that the `title` and `aria-label` together doesn't cause duplication of the label being read, I think that could be a fix. I just tested VoiceOver on Chrome with a [Codepen](https://codepen.io/andylincoln/pen/agzdbv) that does this sort of thing and it only reads the `aria-label` even with both.

I can follow up on how that behaves in NVDA on Windows with other browsers.

**Comment by andylincoln** [comment]
@jaketrent Haven't tested it yet, thanks for reminding me. I'll make a note to test it with NVDA and JAWS now that I have the JAWS laptop in the Boston office on Monday.

**Comment by andylincoln** [comment]
Interesting finding on IE 11 with JAWS 2019 where both the `title` and `aria-label` attributes are used, the `aria-label` is read first, followed by the `title`, which is then visualized as a tooltip via JAWS. Here's a screencap of the result and the code I used:
<img width="497" alt="ie11-jaws-2019-title-label" src="https://user-images.githubusercontent.com/2366108/64185942-ffb5af00-ce3b-11e9-8fc6-7f77f407b017.png">

```html
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="description" content="">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Title Attribute Accessibility Test</title>
</head>
<body>
    <h1>Test Page</h1>
    <button title="Title being used as tooltip" aria-label="This is the aria label"><span aria-hidden="true">
      Accessibly hidden text</span>
    </button>
    <button>Control Test - should always be read as the "label" by default</button>
</body>

</html>```

**Comment by andylincoln** [comment]
Different behavior with NVDA + Firefox using the same file:
On page load, when the whole page is read, the `aria-label` is read and the tooltip is not, as I expected. However, the `aria-label` is not read upon `TAB`ing to the button, instead the `title` is the only thing that is read, so the `aria-label` is ignored in all other interactions.

The conclusion I have here is that a title as tooltip is acceptable as _additional_ information, but not as a replacement for a label.  `<button>` text, `<label>` or `aria-label` should be preferred for establishing the accessible name. I'd have to do more research to see if `title` gets counted as an "accessible description" here as defined by the[ Accessible Name and Description docs](https://www.w3.org/TR/accname-1.1/#dfn-accessible-description).

**Comment by jaketrent** [comment]
@danethurber , I'm missing standup today.  Will you triage this w/ the team, please?

**Comment by danethurber** [comment]
At the moment the `Card.Action` will pass through the `aria-label`. I'll update the props to require an `aria-label` going forward. As we address other components this is a great amount of information to use when ensuring our accessibility support.

As for the issue on the home page, the team responsible will need to update their implementation to add the `aria-label`. I'll make the proptype change and publish a new version and they can update when they can.

**Comment by andylincoln** [comment]
@danethurber Awesome! That sounds like a good plan to me.
